File: tests/examples/java/batch/src/test/java/BatchSmoke.java
import java.sql.DriverManager;
import java.time.Instant;
import java.util.List;
import tech.ydb.common.transaction.TxMode;
import tech.ydb.core.grpc.GrpcTransport;
import tech.ydb.query.QueryClient;
import tech.ydb.query.tools.SessionRetryContext;

/** Run against a disposable database without a books table. */
public final class BatchSmoke {
    private static final String SCHEMA = "CREATE TABLE books (book_id Uint64 NOT NULL, author_id Uint64 NOT NULL, isbn Utf8 NOT NULL, book_type Utf8 NOT NULL, title Utf8 NOT NULL, year Int32 NOT NULL, available Timestamp NOT NULL, tags Json NOT NULL, PRIMARY KEY(book_id));";
    public static void main(String[] args) throws Exception {
        String endpoint = java.util.Objects.requireNonNull(System.getenv("YDB_CONNECTION_STRING"));
        try (var connection = DriverManager.getConnection("jdbc:ydb:" + endpoint)) {
            connection.createStatement().execute(SCHEMA);
            try {
                var jdbc = new batch.jdbc.Queries(connection);
                jdbc.createBooks(List.of());
                check(jdbc.booksByYear(2026).isEmpty());
                jdbc.createBooks(List.of(new batch.jdbc.CreateBooksBooksItem(-1L, 42L, "isbn", "paper", "First", 2026, Instant.EPOCH, "[1,true]"),
                        new batch.jdbc.CreateBooksBooksItem(7L, 42L, "isbn2", "paper", "Second", 2026, Instant.EPOCH, "[]")));
                check(jdbc.booksByYear(2026).size() == 2);
                check(jdbc.booksByYear(2026).stream().anyMatch(row -> row.bookId() == -1L && row.tags().equals("[1,true]")));
                connection.setAutoCommit(false);
                jdbc.createBooks(List.of(new batch.jdbc.CreateBooksBooksItem(8L, 42L, "rollback", "paper", "Rollback", 2027, Instant.EPOCH, "[]")));
                connection.rollback();
                connection.setAutoCommit(true);
                check(jdbc.booksByYear(2027).isEmpty());
                try (var transport = GrpcTransport.forConnectionString(endpoint).build(); var client = QueryClient.newClient(transport).build()) {
                    var retry = SessionRetryContext.create(client).build();
                    retry.supplyResult(session -> {
                        var tx = session.createNewTransaction(TxMode.SERIALIZABLE_RW);
                        var nativeQueries = new batch.nativeapi.Queries(tx);
                        nativeQueries.createBooks(List.of());
                        nativeQueries.createBooks(List.of(new batch.nativeapi.CreateBooksBooksItem(9L, 42L, "native", "paper", "Native", 2028, Instant.EPOCH, "[false]")));
                        check(nativeQueries.booksByYear(2028).get(0).tags().equals("[false]"));
                        return tx.commit();
                    }).join().getValue();
                    check(jdbc.booksByYear(2028).size() == 1);
                    retry.supplyStatus(session -> {
                        var tx = session.createNewTransaction(TxMode.SERIALIZABLE_RW);
                        new batch.nativeapi.Queries(tx).createBooks(List.of(new batch.nativeapi.CreateBooksBooksItem(10L, 42L, "rollback", "paper", "Rollback", 2029, Instant.EPOCH, "[]")));
                        return tx.rollback();
                    }).join().expectSuccess();
                    check(jdbc.booksByYear(2029).isEmpty());
                }
                System.out.println("Java native/JDBC batch smoke passed");
            } finally {
                if (!connection.getAutoCommit()) { connection.rollback(); connection.setAutoCommit(true); }
                connection.createStatement().execute("DROP TABLE books;");
            }
        }
    }
    private static void check(boolean value) { if (!value) throw new AssertionError("batch insert contract failed"); }
}

File: tests/examples/kotlin/src/test/kotlin/BatchSmoke.kt
import java.sql.DriverManager
import java.time.Instant
import org.jetbrains.exposed.v1.core.DatabaseConfig
import org.jetbrains.exposed.v1.jdbc.Database
import tech.ydb.common.transaction.TxMode
import tech.ydb.core.grpc.GrpcTransport
import tech.ydb.query.QueryClient
import tech.ydb.query.tools.SessionRetryContext
import tech.ydb.exposed.dialect.registerYdbDialect
import tech.ydb.exposed.dialect.ydbTransaction

private const val SCHEMA = "CREATE TABLE books (book_id Uint64 NOT NULL, author_id Uint64 NOT NULL, isbn Utf8 NOT NULL, book_type Utf8 NOT NULL, title Utf8 NOT NULL, year Int32 NOT NULL, available Timestamp NOT NULL, tags Json NOT NULL, PRIMARY KEY(book_id));"
private class ExpectedRollback : RuntimeException()

/** Run against a disposable database without a books table. */
fun main() {
    val endpoint = requireNotNull(System.getenv("YDB_CONNECTION_STRING"))
    DriverManager.getConnection("jdbc:ydb:$endpoint").use { connection ->
        connection.createStatement().use { it.execute(SCHEMA) }
        try {
            val jdbc = batch.jdbc.Queries(connection)
            jdbc.createBooks(emptyList())
            check(jdbc.booksByYear(2026).isEmpty())
            jdbc.createBooks(listOf(
                batch.jdbc.CreateBooksBooksItem(-1L, 42, "isbn", "paper", "First", 2026, Instant.EPOCH, "[1,true]"),
                batch.jdbc.CreateBooksBooksItem(7, 42, "isbn2", "paper", "Second", 2026, Instant.EPOCH, "[]")))
            check(jdbc.booksByYear(2026).size == 2)
            check(jdbc.booksByYear(2026).any { it.bookId == -1L && it.tags == "[1,true]" })
            connection.autoCommit = false
            jdbc.createBooks(listOf(batch.jdbc.CreateBooksBooksItem(8, 42, "rollback", "paper", "Rollback", 2027, Instant.EPOCH, "[]")))
            connection.rollback()
            connection.autoCommit = true
            check(jdbc.booksByYear(2027).isEmpty())
            GrpcTransport.forConnectionString(endpoint).build().use { transport ->
                QueryClient.newClient(transport).build().use { client ->
                    val retry = SessionRetryContext.create(client).build()
                    val native = batch.nativeapi.Queries(retry)
                    native.createBooks(emptyList())
                    native.createBooks(listOf(batch.nativeapi.CreateBooksBooksItem(9, 42, "native", "paper", "Native", 2028, Instant.EPOCH, "[false]")))
                    check(jdbc.booksByYear(2028).single().tags == "[false]")
                    retry.supplyStatus { session ->
                        val transaction = session.createNewTransaction(TxMode.SERIALIZABLE_RW)
                        batch.nativeapi.Queries(transaction).createBooks(listOf(batch.nativeapi.CreateBooksBooksItem(10, 42, "rollback", "paper", "Rollback", 2029, Instant.EPOCH, "[]")))
                        transaction.rollback()
                    }.join().expectSuccess()
                    check(jdbc.booksByYear(2029).isEmpty())
                }
            }
            registerYdbDialect()
            val database = Database.connect("jdbc:ydb:$endpoint", driver = "tech.ydb.jdbc.YdbDriver", databaseConfig = DatabaseConfig { useNestedTransactions = false })
            ydbTransaction(database) {
                val exposed = batch.exposed.Queries(this)
                exposed.createBooks(emptyList())
                exposed.createBooks(listOf(batch.exposed.CreateBooksBooksItem(11, 42, "exposed", "paper", "Exposed", 2030, Instant.EPOCH, "[true]")))
                check(jdbc.booksByYear(2030).isEmpty())
            }
            check(jdbc.booksByYear(2030).single().tags == "[true]")
            try {
                ydbTransaction(database) {
                    batch.exposed.Queries(this).createBooks(listOf(batch.exposed.CreateBooksBooksItem(12, 42, "rollback", "paper", "Rollback", 2031, Instant.EPOCH, "[]")))
                    throw ExpectedRollback()
                }
            } catch (_: ExpectedRollback) { }
            check(jdbc.booksByYear(2031).isEmpty())
            println("Kotlin native/JDBC/Exposed batch smoke passed")
        } finally {
            if (!connection.autoCommit) { connection.rollback(); connection.autoCommit = true }
            connection.createStatement().use { it.execute("DROP TABLE books;") }
        }
    }
}
